File: FirstAkkaApplication/src/main/scala/akka/first/app/scala/actors/AggregateActor.scala
package akka.first.app.scala.actors

import scala.collection.immutable._

import akka.actor.Actor
import akka.first.app.scala.ReduceData
import akka.first.app.scala.Result

class AggregateActor extends Actor {

  var finalReducedMap = new HashMap[String, Int]

  def receive: Receive = {
    case message: ReduceData =>
      aggregateInMemoryReduce(message.reduceDataMap)
    case message: Result =>
      println(finalReducedMap.toString())
  }

  def aggregateInMemoryReduce(reducedMap: Map[String, Int]) {
    var count: Int = 0
    reducedMap.foreach((entry: (String, Int)) =>
      if (finalReducedMap.contains(entry._1)) {
        count = entry._2 + finalReducedMap.get(entry._1).get
        finalReducedMap += entry._1 -> count
      } else
        finalReducedMap += entry._1 -> entry._2)
  }
}

File: FirstAkkaApplication/src/main/java/akka/first/app/java/actors/AggregateActor.java
package akka.first.app.java.actors;

import java.util.HashMap;
import java.util.Map;

import akka.actor.UntypedActor;
import akka.first.app.java.messages.ReduceData;
import akka.first.app.java.messages.Result;

public class AggregateActor extends UntypedActor {

	private Map<String, Integer> finalReducedMap = new HashMap<String, Integer>();

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof ReduceData) {
			ReduceData reduceData = (ReduceData) message;
			aggregateInMemoryReduce(reduceData.getReduceDataList());
		} else if (message instanceof Result) {
			System.out.println(finalReducedMap.toString());
		} else
			unhandled(message);
	}

	private void aggregateInMemoryReduce(Map<String, Integer> reducedList) {
		Integer count = null;
		for (String key : reducedList.keySet()) {
			if (finalReducedMap.containsKey(key)) {
				count = reducedList.get(key) + finalReducedMap.get(key);
				finalReducedMap.put(key, count);
			} else {
				finalReducedMap.put(key, reducedList.get(key));
			}
		}
	}
}
